onnxruntime/core/providers/cuda/cuda_mempool_arena.cc

```diff
@@ -0,0 +1,239 @@
+// Copyright (c) Microsoft.
+// Licensed under the MIT License.
+
+#include "cuda_mempool_arena.h"
+
+#include <algorithm>
+
+#include "core/providers/cuda/shared_inc/cuda_call.h"  // ORT CudaCall helpers
+#include "core/providers/shared_library/provider_api.h"
+
+namespace onnxruntime {
+namespace cuda {
+
+// ======== CudaMempoolArena ========
+
+CudaMempoolArena::CudaMempoolArena(const OrtMemoryInfo& memory_info,
+                                   uint64_t pool_release_threshold,
+                                   size_t bytes_to_keep_on_shrink,
+                                   const logging::Logger* logger)
+    : IArena(memory_info),
+      pool_release_threshold_(pool_release_threshold),
+      bytes_to_keep_on_shrink_(bytes_to_keep_on_shrink),
+      logger_(logger) {
+  if (logger_ == nullptr) {
+    logger_ = &::onnxruntime::logging::LoggingManager::DefaultLogger();
+  }
+
+  // Create a process-local device memory pool for device_id_.
+  // 'cudaMemAllocationTypeDevice' (for cudaMemPoolProps.allocType) not clear when it is available
+
+  cudaMemPoolProps props{};
+  // Pinned is not the same as pinned allocator, cudaMemLocationTypeDevice actually does not exist
+  // even though is present in some internet docs.
+  props.allocType = cudaMemAllocationTypePinned;
+  props.handleTypes = cudaMemHandleTypeNone;        // local to process
+  props.location.type = cudaMemLocationTypeDevice;  // Device memory
+  props.location.id = this->Info().device.Id();
+
+  CUDA_CALL_THROW(cudaMemPoolCreate(&pool_, &props));
+
+  if (pool_release_threshold_ != 0) {
+    CUDA_CALL_THROW(cudaMemPoolSetAttribute(pool_, cudaMemPoolAttrReleaseThreshold,
+                                            &pool_release_threshold_));
+  }
+
+  LOGS(*logger_, INFO) << "CudaMempoolArena created on device " << this->Info().device.Id()
+                       << " with pool_release_threshold=" << pool_release_threshold_
+                       << " bytes_to_keep_on_shrink=" << bytes_to_keep_on_shrink_ << ".";
+
+  // Intentionally DO NOT call cudaDeviceSetMemPool(device_id_, pool_);
+  // All allocations explicitly target this pool via cudaMallocFromPoolAsync.
+}
+
+CudaMempoolArena::~CudaMempoolArena() {
+  // 1) Best-effort: enqueue frees for any remaining allocations on their recorded streams.
+  //    No locking by design: destruction implies no concurrent access.
+  for (auto& kv : alloc_map_) {
+    void* p = kv.first;
+    const cudaStream_t s = kv.second.stream;
+    (void)cudaFreeAsync(p, s);  // ignore errors in destructor
+  }
+
+  // Now it is safe to drop our bookkeeping.
+  alloc_map_.clear();
```

Comment:
We clear `stream_map_` here but it is actually used in the next operation - `SyncAllKnownStreams_NoThrow()` ?